protocol/x/clob/keeper/deleveraging.go

```diff
@@ -188,6 +188,56 @@ func (k Keeper) CanDeleverageSubaccount(
 	return false, clobPair.Status == types.ClobPair_STATUS_FINAL_SETTLEMENT, nil
 }
 
+// GateWithdrawalsIfNegativeTncSubaccountSeen gates withdrawals if a negative TNC subaccount exists.
+// It does this by inserting a zero-fill deleveraging operation into the operations queue iff any of
+// the provided negative TNC subaccounts are still negative TNC.
+func (k Keeper) GateWithdrawalsIfNegativeTncSubaccountSeen(
+	ctx sdk.Context,
+	negativeTncSubaccountIds []satypes.SubaccountId,
+) (err error) {
+	foundNegativeTncSubaccount := false
+	var negativeTncSubaccountId satypes.SubaccountId
+	for _, subaccountId := range negativeTncSubaccountIds {
+		bigNetCollateral,
+			_,
+			_,
+			err := k.subaccountsKeeper.GetNetCollateralAndMarginRequirements(
+			ctx,
+			satypes.Update{SubaccountId: subaccountId},
+		)
+		if err != nil {
+			return err
+		}
+
+		// If the subaccount has negative TNC, mark that a negative TNC subaccount was found.
+		if bigNetCollateral.Sign() == -1 {
+			foundNegativeTncSubaccount = true
+			negativeTncSubaccountId = subaccountId
+			break
+		}
+	}
+
+	if !foundNegativeTncSubaccount {
+		return nil
+	}
+
+	// A negative TNC subaccount was found, therefore insert a zero-fill deleveraging operation into
+	// the operations queue to indicate withdrawals should be gated.
+	subaccount := k.subaccountsKeeper.GetSubaccount(ctx, negativeTncSubaccountId)
+	perpetualPositions := subaccount.GetPerpetualPositions()
+	if len(perpetualPositions) == 0 {
+		return errorsmod.Wrapf(
+			types.ErrNoPerpetualPositionsToLiquidate,
+			"GateWithdrawalsIfNegativeTncSubaccountSeen: subaccount has no open positions: (%s)",
+			lib.MaybeGetJsonString(subaccount),
+		)
+	}
+	perpetualId := subaccount.PerpetualPositions[0].PerpetualId
```

Comment:
Is there any case a subaccount can have negative TNC and have a no perpetual positions? ie price slips fast enough and we match/liquidate at a less favorable price than the bankruptcy price?

In this case do we want to default to some value if this doesn't exist? I don't think the perpetual id is used for zero fill deleveraging operations, so having a default would be good to protect against the error when we access a nonexisting array element.

Reply:
> Is there any case a subaccount can have negative TNC and have a no perpetual positions? ie price slips fast enough and we match/liquidate at a less favorable price than the bankruptcy price?

shouldn't be possible since subaccounts shouldn't be able to close out all of their positions and still have negative tnc

if this was possible, it would require a subaccount state transition to occur at a worse price than subaccount's bankruptcy price (price that subaccount tnc goes to zero), and that should be invalid

i agree it's still good to think through this though in case invariant gets violated. want me to create an m7 ticket to handle this? also maybe worth discussing with rest of CLOB team to ensure it never happens

subaccounts shouldn't be able to close out all of their positions and still have negative tnc. it shouldn't be possible to close a position and still have negative TNC also note subaccounts [aren't even liquidatable if they have zero positions](https://github.com/dydxprotocol/v4-chain/blob/6014ec261478d509e1ffa6742b40fa5967e970a9/protocol/x/clob/keeper/liquidations.go#L376), so i guess that means they wouldn't be deleverage-able either? so if it's possible for subaccounts to be in this state, then i dont think theres any way out of it (besides transferring the subac

Reply:
i think quick fix we should just add a defeault value of 0, and make a ticket to think through edge cases later

Reply:
synced on slack and we're not doing this for simplicity. created ticket here: https://linear.app/dydx/issue/CLOB-1061/use-default-perpetual-id-for-withdrawal-gating